### PR TITLE
[ML] Randomize train/test cluster boundary assignment in RDataLoader

### LINKED ISSUE #22194: RDataLoader results in same train/test split across runs
### Explain what you would like to see improved and how.

When using the `RDataLoader` for splitting a dataset into train/test subsets, the first fraction of every cluster is always assigned to training and the last one is always assigned to testing, this does not pose a problem within one complete training loop over N epochs, but this does in fact mean that across different runs, the resulting train/test datasets will always be the same.

### ROOT version

v6.40-rc

### Installation method

all

### Operating system

all

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by siliataider**
# This Pull request:

Previously `RDataLoader` always assigned the first fraction of each cluster to training and the last fraction to validation. This meant that across different runs, the train/val split was always identical regardless of the seed.

This PR fixes the issue by using the shuffle seed to randomly decide, per cluster, whether training takes the prefix or suffix of that cluster.

This PR fixes #22194

### CODE REVIEW

File: bindings/pyroot/pythonizations/test/ml_dataloader.py [hunk 1]
```diff
@@ -4570,6 +4570,45 @@ def test16_vector_padding(self):
...
+
+            val_1_collected = sorted([v for x, y in gen_val1.as_numpy() for v in x.flatten().tolist()])
+            val_2_collected = sorted([v for x, y in gen_val2.as_numpy() for v in x.flatten().tolist()])
+
+            self.assertNotEqual(val_1_collected, val_2_collected)
+
+        except:
+            self.teardown_file(self.file_name1)
+            self.teardown_file(self.file_name2)
+            raise
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/test/ml_dataloader.py]: I would rather use a `finally` block here which wouldn't need to deal with re-raising the exception

File: tree/ml/inc/ROOT/ML/RClusterLoader.hxx [hunk 2]
```diff
@@ -227,18 +227,27 @@ public:
          // --- Shuffled path
          // Every cluster contributes a prefix to training and a suffix to validation.
          // Cost: Each cluster is read twice per epoch, only when validation split is more than 0.
-         // TODO(staider) Swicth between prefix or suffix for validation randomly per cluster
+         // We generate a random boolean value to decide whether the training set gets the prefix
+         // or suffix of each cluster to ensure better shuffling across runs when splitting.
+         std::mt19937 g(fSetSeed);
+         std::uniform_int_distribution<int> coin(0, 1);
```
Comment (vepadulano) [inline tree/ml/inc/ROOT/ML/RClusterLoader.hxx:233]: nitpick, as a micro-optimization we could use
```suggestion
         std::uniform_int_distribution<std::uint8_t> coin(0, 1);
```
And then below
```
bool trainIsPrefix = coin(g);
```

**Review by vepadulano — APPROVED** [review]
Nice! See minor considerations from my side.

**Comment by dpiparo** [comment]
Is a backport to 6.40 needed?

File: tree/ml/inc/ROOT/ML/RClusterLoader.hxx (hunk 2 above)
Comment (siliataider) [reply tree/ml/inc/ROOT/ML/RClusterLoader.hxx:233]: Had to revert back to `int` for windows to build:

> invalid template argument for uniform_int_distribution: N4950 [rand.req.genl]/1.5 requires one of short, int, long, long long, unsigned short, unsigned int, unsigned long, or unsigned long long [C:\ROOT-CI\build\tree\ml\G__ROOTMLDataLoader.vcxproj]
